### ISSUE TITLE
Unresolvable "Raft log too large" state

### ISSUE THREAD

**Issue by dankinder** [cockroachdb/cockroach#60538]
**Describe the problem**

I have a cluster in a strange state and I've posted on the forum about it as well, see [here](https://forum.cockroachlabs.com/t/under-replicated-ranges-after-decommission/4239/4) and [here](https://forum.cockroachlabs.com/t/problem-ranges-report-what-can-be-done-about-it/2391/7)

There are 2 issues:

1) Some ranges have 2 replicas together on the same node, which makes decommission never finish. [This forum post](https://forum.cockroachlabs.com/t/under-replicated-ranges-after-decommission/4239/3) references those.

2) One range is saying "Raft log too large" and is never getting out of that state, despite trying to run it through several different queues. Here are screenshots of its state.

I was hoping to mitigate this by backing up the cluster and restoring from backup, but *the backup job just sits at 0%*. IMPORT also hangs. I assume because of weirdness with this range.

![Screen Shot 2021-02-12 at 2 23 56 PM](https://user-images.githubusercontent.com/5198575/107813880-02117d80-6d3f-11eb-8897-5081c7809baa.png)
![Screen Shot 2021-02-12 at 2 24 07 PM](https://user-images.githubusercontent.com/5198575/107813881-02aa1400-6d3f-11eb-9154-fe5e0746e0c0.png)
![Screen Shot 2021-02-12 at 2 24 20 PM](https://user-images.githubusercontent.com/5198575/107813882-02aa1400-6d3f-11eb-9831-730aa5da7679.png)
![Screen Shot 2021-02-12 at 2 24 28 PM](https://user-images.githubusercontent.com/5198575/107813883-02aa1400-6d3f-11eb-9765-82ccbeeacc07.png)
![Screen Shot 2021-02-12 at 2 24 44 PM](https://user-images.githubusercontent.com/5198575/107813884-02aa1400-6d3f-11eb-9263-e25816138d45.png)

**To Reproduce**

I was inserting as fast as I could into this cluster, batched into 100 rows per transaction (plus some regular IMPORT before that). It's a bulk import and we hope to use the cluster in production once this is done.

I have a debug zip I can upload if you want it.

I've scoured the github issues and am having difficulty finding any way to get out of this state, and I suspect there should be a way written down somewhere. I tried increasing the max_rate for recovery and rebalance at the suggestion of an issue somewhere but that didn't seem to help.

**Environment:**
 - CockroachDB version 2.2.3
 - Server OS: Centos 7

**Comment by nvanbenschoten** [comment]
Hi @dankinder,

Thanks for the report. A `debug.zip` would be helpful in diagnosing what's going wrong with this range. One thing we see is that each replica has a long uncommitted tail in its raft log. We also see that the leadership term is quite high. So one theory is that the range is thrashing leaders and each leader is getting stuck probing the followers' raft logs and never completing by the time it loses its leadership.

To detect this, we could look for two things:
1. do you see the leader on this range changing?
2. could you visit `http://<host of current leader>:8080/debug/logspy?count=1000&duration=10s&grep=r191` and post the output? If you hit an error, you may need to run `SET CLUSTER SETTING server.remote_debugging.mode = 'any';`.

**Comment by nvanbenschoten** [comment]
Also, what kind of network latency do you expect between the peers in this range? Is this a multi-region cluster?

**Comment by nvanbenschoten** [comment]
@dankinder do you mind opening a second github issue to discuss the under-replicated range issue mentioned in https://forum.cockroachlabs.com/t/under-replicated-ranges-after-decommission/4239. It seems like a separate problem that will have a separate resolution.

**Comment by dankinder** [comment]
Thanks @nvanbenschoten! 

I'll try to get you that debug zip soon but here's a quick dump from the logs grepping for r191:

```
W210212 21:36:13.433690 1865342 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,consistencyChecker,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
W210212 21:44:57.311519 1902739 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 21:44:57.372690 1902739 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.061215711s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
I210212 21:46:23.432675 1865342 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,consistencyChecker,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 11m9.999204063s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
I210212 21:46:23.510374 1818311 kv/kvserver/replica_write.go:181 ⋮ [n6,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow command ‹RequestLease [/Table/54/1/"keio.ac.jp"/"www.ht.sfc"/"/~mics/mdp/pukiwiki/index.php"/"cmd=diff&page=%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5Bcmd%3Dbackup%26page%3D%5B%5B%5B%5B%5B%5Bcmd%3Dbackup%26page%3D%255B%255B%255B%255B%255B%255B%255B%255Bcmd%253Dbackup%2526page%253D%255B%255B%255B%255B%255B%255B%255B%255Bcmd%253Ddiff%2526page%253D%255B%255B%255B%255Bcmd%253Ddiff%2526page%253D%255B%255Bpage%253D%255B%255B%255B%255B%25A5%25D8%25A5%25EB%25A5%25D7%255D%255D%253D%255D%255D%2526pcmd%253Dupload%2526plugin%253Dattach%255D%255D%255D%255D%253D%255D%255D%255D%255D%253D%255D%255D%253D%255D%255D%253D%255D%255D%255D%255D%253D%255D%255D%253D%255D%255D%253D%255D%255D%5D%5D%3D%5D%5D%3D%5D%5D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D"/"https"/1970-01-01T00:00:00Z,/Min)› finished after 1340.25s with error ‹result is ambiguous (context canceled)›
W210212 21:47:23.510630 1912859 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,consistencyChecker,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
E210212 21:47:23.628198 1912916 kv/kvserver/replica_write.go:222 ⋮ [n6,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] range unavailable: ‹have been waiting 60.00s for proposing command RequestLease [/Table/54/1/"keio.ac.jp"/"www.ht.sfc"/"/~mics/mdp/pukiwiki/index.php"/"cmd=diff&page=%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5B%5Bcmd%3Dbackup%26page%3D%5B%5B%5B%5B%5B%5Bcmd%3Dbackup%26page%3D%255B%255B%255B%255B%255B%255B%255B%255Bcmd%253Dbackup%2526page%253D%255B%255B%255B%255B%255B%255B%255B%255Bcmd%253Ddiff%2526page%253D%255B%255B%255B%255Bcmd%253Ddiff%2526page%253D%255B%255Bpage%253D%255B%255B%255B%255B%25A5%25D8%25A5%25EB%25A5%25D7%255D%255D%253D%255D%255D%2526pcmd%253Dupload%2526plugin%253Dattach%255D%255D%255D%255D%253D%255D%255D%255D%255D%253D%255D%255D%253D%255D%255D%253D%255D%255D%255D%255D%253D%255D%255D%253D%255D%255D%253D%255D%255D%5D%5D%3D%5D%5D%3D%5D%5D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D%3D%5D%5D"/"https"/1970-01-01T00:00:00Z,/Min).›
‹Descriptor:  r191:{-} [(n6,s18):279, (n5,s11):287, (n47,s96):285, (n23,s72):289LEARNER, next=290, gen=1300, sticky=1612382626.811532473,0]›
W210212 21:51:37.710660 1931788 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 21:51:37.710950 1931788 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.000445752s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
W210212 21:52:37.711365 1935187 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 21:52:40.176462 1935187 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m2.46520458s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
W210212 21:53:40.176816 1940727 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 21:53:40.177141 1940727 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.000566499s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
W210212 21:54:40.177486 1945445 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 21:54:40.177835 1945445 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.000455019s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
W210212 21:55:40.178280 1949867 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 21:55:40.253098 1949867 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.074901793s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
I210212 21:57:33.579494 1912859 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,consistencyChecker,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 11m10.069024048s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
W210212 21:58:33.696489 1962706 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,consistencyChecker,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
W210212 22:02:46.311169 1982479 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 22:02:46.393280 1982479 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.082201361s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
I210212 22:08:43.703577 1962706 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,consistencyChecker,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 11m10.007300517s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
W210212 22:09:19.311495 2010543 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 22:09:19.346256 2010543 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.034836972s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
W210212 22:09:43.986355 2012279 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,consistencyChecker,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
W210212 22:15:44.872400 2037073 kv/kvserver/replica_range_lease.go:1064 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] have been waiting 1m0s attempting to acquire lease
I210212 22:15:44.908799 2037073 kv/kvserver/replica_range_lease.go:1069 ⋮ [n6,replicate,s18,r191/279:‹/Table/54/1/"ke{io.ac…-k.jp"…}›] slow lease acquisition finished after 1m0.036505932s with error ‹[NotLeaseHolderError] r191: replica (n6,s18):279 not lease holder; lease holder unknown› after 1 attempts
```

**Comment by dankinder** [comment]
And I don't see the leader changing, it's always n6

**Comment by dankinder** [comment]
K I just uploaded a debug zip to the support portal.

And actually the leader does seem to change between n5 and n6, it was just hard to catch it, so your theory may be correct.

This cluster is mostly in one datacenter but one node is in another datacenter. Actually that's part of what I was trying to do, migrate it from one set of nodes to another, and that migration couldn't complete because of this issue. So I had to recommission the nodes and leave everything in the cluster.

**Comment by nvanbenschoten** [comment]
Hi @dankinder, thanks for sending over the debug zip. I was able to confirm what you're seeing.

Regarding the first issue, the team was able to use the information you provided in the forum post to piece together what was going wrong in https://github.com/cockroachdb/cockroach/issues/60545. They landed a fix on `master` and have backported it to our 20.2 release branch, so it should be included in the next 20.2 point release.

Regarding the second issue, this is something that we've seen one other time. In the other case and it sounds like in this one, we saw the uncommitted tail of the raft log of a range diverge across two potential leaders in response to a large influx of write traffic. The range then struggled to reconcile this divergence and was unable to hold a steady leadership term long enough to recover. We recently landed https://github.com/etcd-io/etcd/pull/11964 in the Raft library we share with etcd, which makes this reconciliation process significantly (in the other case, ~1000x) more efficient, which would allow the range to get out of this bad state within seconds. There's a bit of uncertainty about why Raft leadership is unstable, but I fear that this will be difficult to diagnose on this cluster.

So there are a few options here for recovering this range. One is to upgrade these nodes to a binary built off of CRDB's `master`, which will have fixes for both issues. However, I understand that you may not want to do that. So another option is to upgrade these nodes to a binary build off of CRDB's `release-20.2` branch to pick up the fix for the first issue. Then, to address the second issue, you could use configure Raft election timeouts to be slightly less aggressive for the time being by starting these processes with the `COCKROACH_RAFT_ELECTION_TIMEOUT_TICKS=45` environment variable. That should help stabilize the leader on this range until it can properly recover.

**Comment by dankinder** [comment]
Got it, thanks for the info and for looking into it. Since we didn't find a workaround too quickly I ended up restoring this data from a backup again (thankfully in this particular instance that was an option), but this will be super useful if/when we encounter these issues again.

**Comment by nvanbenschoten** [comment]
Thanks Dan, I really appreciate your help with this issue. I'm going to close this for now, since it sounds like the two issues have resolutions that are merged into master and that we'll be looking to push back into past releases, but please ping this thread if anything else comes up on your end.
